app/services/media_storage/azure_adapter.rb

```diff
@@ -0,0 +1,92 @@
+# frozen_string_literal: true
+
+module MediaStorage
+  class AzureAdapter < AbstractAdapter
+    attr_reader :prefix, :public_container, :private_container, :storage_account_name, :get_expiration, :put_expiration, :client, :signer
+    DEFAULT_EXPIRES_IN = 3 # time in minutes, see get_expiry_time(expires_in)
+
+    def initialize(opts={})
+      @storage_account_name = opts[:azure_storage_account]
+      @public_container = opts[:azure_storage_container_public]
+      @private_container = opts[:azure_storage_container_private]
+      @prefix = opts[:prefix] || Rails.env
+      @get_expiration = opts.dig(:expiration, :get) || DEFAULT_EXPIRES_IN
+      @put_expiration = opts.dig(:expiration, :put) || DEFAULT_EXPIRES_IN
+
+      @client = Azure::Storage::Blob::BlobService.create(
+        storage_account_name: opts[:azure_storage_account],
+        storage_access_key: opts[:azure_storage_access_key]
+      )
+      @signer = Azure::Storage::Common::Core::Auth::SharedAccessSignature.new(
+        opts[:azure_storage_account],
+        opts[:azure_storage_access_key]
+      )
+    end
+
+    def stored_path(content_type, medium_type, *path_prefix)
+      extension = get_extension(content_type)
+      path = prefix.to_s
+      path += '/' unless path[-1] == '/'
+      path += "#{medium_type}/"
+      path += "#{path_prefix.join('/')}/" unless path_prefix.empty?
+      path + "#{SecureRandom.uuid}.#{extension}"
+    end
+
+    def get_path(path, opts={})
+      if opts[:private]
+        signer.signed_uri(
+          client.generate_uri("#{private_container}/#{path}"),
+          false,
+          service: 'b', # blob
+          permissions: 'r', # read
+          expiry: get_expiry_time(opts[:get_expires] || get_expiration)
+        ).to_s
+      else
+        "https://#{path}"
+      end
+    end
+
+    def put_path(path, opts={})
+      container = opts[:private] ? private_container : public_container
+
+      signer.signed_uri(
+        client.generate_uri("#{container}/#{path}"),
+        false,
+        service: 'b', # blob
+        permissions: 'rcw', # read create write
```

Comment:
Not blocking - Folks need the 'r' read permission to put the file?

Reply:
Ah true, they really don't.